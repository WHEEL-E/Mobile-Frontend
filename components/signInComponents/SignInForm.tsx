import React from "react";
import { useTranslation } from "react-i18next";
import { ScrollView, StyleSheet, View } from "react-native";
import {
  Field,
  InjectedFormProps,
  reduxForm,
  WrappedFieldProps,
} from "redux-form";
import colors from "../../utilities/constants/colors";
import { PADDING_VERTICAL } from "../../utilities/constants/spacing";
import {
  validateMail,
  validateNotEmpty,
  validatePassword,
} from "../../utilities/dataValidators";
import { SignInData, submitLoginForm } from "../../utilities/types/signInTypes";
import { RoundEdgedButton } from "../buttons/RoundEdgedButton";
import PickerComponent from "../signUpComponents/Picker";
import { RenderInputComponent } from "../signUpComponents/RenderInputComponent";

const SignInForm = (props: InjectedFormProps<SignInData>) => {
  const typeLabels = ["supervisor", "patient"];
  const { handleSubmit } = props;
  const { t } = useTranslation();

  return (
    <View style={styles.container}>
      <ScrollView
        contentContainerStyle={styles.scrollView}
        nestedScrollEnabled={true}
      >
        <Field
          name="emailAddress"
          component={RenderInputComponent}
          validate={validateMail}
          warn={validateMail}
        />
        <Field
          name="password"
          component={RenderInputComponent}
          validate={validatePassword}
          warn={validatePassword}
        />
        <Field
          name="type"
          component={(props: WrappedFieldProps) => (
            <PickerComponent fieldProps={props} labels={typeLabels} />
          )}
          validate={validateNotEmpty}
          warn={validateNotEmpty}
        />
      </ScrollView>
      <RoundEdgedButton
        title={t("signInScreen.logIn")}
        onPress={handleSubmit(submitLoginForm)}
        backgroundColor={colors.darkGreen}
      />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    width: "100%",
  },
  scrollView: {
    justifyContent: "center",
    alignItems: "center",
    paddingVertical: PADDING_VERTICAL,
    flex: 1,
  },
});

const Form = reduxForm<SignInData>({
  form: "SignInform",
})(SignInForm);

export default Form;
