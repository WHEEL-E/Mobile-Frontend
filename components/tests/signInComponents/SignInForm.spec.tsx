import React from "react";
import ReduxThunk from "redux-thunk";
import renderer from "react-test-renderer";
import { Provider } from "react-redux";
import { configureStore } from "@reduxjs/toolkit";
import { fireEvent, render } from "@testing-library/react-native";
import reducer from "../../../store/reducers/rootReducer";
import SignInForm from "../../signInComponents/SignInForm";

describe("SignInForm", () => {
  const store = configureStore({ reducer, middleware: [ReduxThunk] });

  const Component = (
    <Provider store={store}>
      <SignInForm />
    </Provider>
  );

  it("Make sure form matches snapshot", () => {
    const signInForm = renderer.create(Component).toJSON();
    expect(signInForm).toMatchSnapshot();
  });

  it("Make sure error message is normally shown", () => {
    const { getAllByTestId } = render(Component);

    const warningText = getAllByTestId("warningText");
    expect(warningText.length).toBe(3);
  });

  it("Make sure error message disappears when adding proper input", () => {
    const signInData = {
      mail: "placeHolder@mail.com",
      password: "Rand0mPassw0rd!",
    };

    const { getByTestId, getAllByTestId } = render(Component);

    const mailField = getByTestId("signInemailAddress");
    const password = getByTestId("signInpassword");

    fireEvent.changeText(mailField, signInData.mail);
    fireEvent.changeText(password, signInData.password);

    const warningText = getAllByTestId("warningText");
    expect(warningText.length).toBe(1);
  });
});
