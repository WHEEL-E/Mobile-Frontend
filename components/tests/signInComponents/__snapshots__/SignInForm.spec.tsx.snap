// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignInForm Make sure form matches snapshot 1`] = `
<View
  style={
    Object {
      "width": "100%",
    }
  }
>
  <RCTScrollView
    contentContainerStyle={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "paddingVertical": 26.68,
      }
    }
  >
    <View>
      <View
        style={
          Object {
            "alignItems": "center",
            "justifyContent": "center",
            "width": "100%",
          }
        }
      >
        <TextInput
          autoComplete="name"
          onChangeText={[Function]}
          placeholder="form.emailAddress"
          style={
            Object {
              "backgroundColor": "#F2F3F7",
              "borderRadius": 15,
              "fontFamily": "Cairo-Medium",
              "fontSize": 26.68,
              "height": 70,
              "marginVertical": 10.672,
              "padding": 26.68,
              "width": "80%",
            }
          }
          testID="signInemailAddress"
          value=""
        />
        <Text
          style={
            Object {
              "color": "red",
              "fontFamily": "Cairo-Regular",
              "fontSize": 20.009999999999998,
              "textAlign": "center",
              "width": "80%",
            }
          }
          testID="warningText"
        >
          signUpScreen.warningMail
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "justifyContent": "center",
            "width": "100%",
          }
        }
      >
        <TextInput
          autoComplete="name"
          onChangeText={[Function]}
          placeholder="form.password"
          style={
            Object {
              "backgroundColor": "#F2F3F7",
              "borderRadius": 15,
              "fontFamily": "Cairo-Medium",
              "fontSize": 26.68,
              "height": 70,
              "marginVertical": 10.672,
              "padding": 26.68,
              "width": "80%",
            }
          }
          testID="signInpassword"
          value=""
        />
        <Text
          style={
            Object {
              "color": "red",
              "fontFamily": "Cairo-Regular",
              "fontSize": 20.009999999999998,
              "textAlign": "center",
              "width": "80%",
            }
          }
          testID="warningText"
        >
          signUpScreen.warningPassword
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "width": "100%",
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "width": "100%",
              },
              Object {
                "zIndex": 5000,
              },
              Object {},
            ]
          }
        >
          <View
            accessibilityState={
              Object {
                "disabled": false,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            forwardedRef={[Function]}
            onClick={[Function]}
            onLayout={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "alignSelf": "center",
                "backgroundColor": "#F2F3F7",
                "borderColor": "#000",
                "borderRadius": 15,
                "borderWidth": 0,
                "flexDirection": "row",
                "height": 70,
                "justifyContent": "space-between",
                "minHeight": 50,
                "opacity": 1,
                "paddingHorizontal": 10,
                "paddingVertical": 3,
                "width": "80%",
                "zIndex": 0,
              }
            }
            testID="signIntype"
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#000",
                    "flex": 1,
                  },
                  Object {
                    "fontFamily": "Cairo-Medium",
                    "fontSize": 26.68,
                  },
                  false,
                  Object {},
                ]
              }
            >
              form.type
            </Text>
            <View
              style={
                Array [
                  Object {
                    "marginLeft": 10,
                  },
                  Object {},
                ]
              }
            >
              <Image
                source={1}
                style={
                  Array [
                    Object {
                      "height": 20,
                      "width": 20,
                    },
                    Object {},
                  ]
                }
              />
            </View>
          </View>
        </View>
        <Text
          style={
            Object {
              "color": "red",
              "fontFamily": "Cairo-Regular",
              "fontSize": 20.009999999999998,
            }
          }
          testID="warningText"
        >
          signUpScreen.warningText
        </Text>
      </View>
    </View>
  </RCTScrollView>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "#11698E",
        "borderColor": "white",
        "borderRadius": 30,
        "borderWidth": 2,
        "height": 55,
        "justifyContent": "center",
        "opacity": 1,
        "width": "80%",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "white",
          "fontFamily": "Cairo-Bold",
          "fontSize": 37.5,
        }
      }
    >
      signInScreen.logIn
    </Text>
  </View>
</View>
`;
